xcode/Subconscious/Shared/Services/DatabaseService.swift

```diff
@@ -85,21 +95,90 @@ final class DatabaseService {
         }
     }
 
+    /// Deletes and rebuilds the database by applying migrations to a new
+    /// database.
+    func rebuild() throws -> Int {
+        try self.database.delete()
+        self.state = .initial
+        do {
+            let version = try self.migrations.migrate(self.database)
+            // If migration succeeded, mark state ready
+            self.state = .ready
+            return version
+        } catch {
+            // Mark state broken by default
+            self.state = .broken
+            throw error
+        }
+    }
+
+    /// Read database metadata from string key
+    func readMetadata(key: String) throws -> String {
+        guard self.state == .ready else {
+            throw DatabaseServiceError.notReady
+        }
+        let rows = try database.execute(
+            sql: "SELECT value FROM database_metadata WHERE key = ?",
+            parameters: [.text(key)]
+        )
+        guard let version = rows.first?.col(0)?.toString() else {
+            throw DatabaseServiceError.notFound(
+                "No value found for key \(key)"
+            )
+        }
+        return version
+    }
+
+    /// Read metadata from type-safe metadata key
+    func readMetadata(key: DatabaseMetaKeys) throws -> String {
+        try readMetadata(key: key.rawValue)
+    }
+
+    /// Write database metadata at string key
+    func writeMetadatadata(key: String, value: String) throws {
+        guard self.state == .ready else {
+            throw DatabaseServiceError.notReady
+        }
+        try database.execute(
+            sql: """
+            INSERT OR REPLACE INTO database_metadata (
+                key,
+                value
+            )
+            VALUES (?, ?)
+            """,
+            parameters: [.text(key), .text(value)]
+        )
+    }
+
+    /// Write database metadata at type-safe metadata key
+    func writeMetadatadata(key: DatabaseMetaKeys, value: String) throws {
+        try writeMetadatadata(key: key.rawValue, value: value)
+    }
+
     /// Write entry syncronously
-    func writeEntry(
-        entry: MemoEntry,
-        info: FileInfo
+    func writeMemo(
```

Comment:
Preferring more direct "write memo to address" signatures wherever possible, to match the shape of the Noosphere API.